Describe how to generate and output secrets (#19886)
github/docs: content/actions/using-workflows/workflow-commands-for-github-actions.md

{% bash %}

{% raw %}
```yaml{:copy}
      - name: Create annotation for build error
        run: echo "::error file=app.js,line=1::Missing semicolon"
```

> {% endraw %}

Author: This was resulting in the `yaml{:copy}` part being rendered as raw....

{% endpowershell %}

### Example: Masking a generated output within a single job

> **Note**

> > The order matters here, you *must* use `add-mask` before you use `set-output`.

Reviewer: - - - - - - - [x] # @@@@@@![IMG20221025114230.jpg](https://user-images.githubusercontent.com/115123317/197852476-6395192c-d479-4006-b6e1-5da943db1c31.jpg)[]()_****_

> > The order matters here, you *must* use `add-mask` before you use `set-output`.

Reviewer 2: Jajaja la 121219 se la cura de mi

As long as you don't need to pass your secret from one job to another job, you can:
1. Generate the secret (without outputting it).
2. Tag it with `add-mask`.
3. Use `set-output` to make it available to other steps within your job.

> {% bash %}

Author: I really don't speak PowerShell (please ignore the fact that I'm a PowerShell contributor). But, yeah, I can figure it out...

Worse, `set-output` was deprecated this month which means all of the examples will need to be updated.

```yaml{:copy}
on: push
jobs:
  generate-a-secret-output:
    runs-on: ubuntu-latest
    steps:
      - id: sets-a-secret
        name: Generate, mask, and output a secret
        run: |
          the_secret=$((RANDOM))
          echo "::add-mask::$the_secret"
          echo "::set-output name=secret-number::$the_secret"
      - name: Use that secret output (protected by a mask)
        run: |
          echo "the secret number is ${{ steps.sets-a-secret.outputs.secret-number }}"
```

{% endbash %}

### Scenario: Passing a secret between jobs or workflows

#### Setup

> 1. Set up a secret-store to be responsible for holding secrets (e.g. Vault).

Reviewer 3: I haven't tested this out, but is using the [GitHub API to create secrets](https://docs.github.com/en/rest/actions/secrets#create-or-update-a-repository-secret) not an option? I don't recall whether you can retrieve a secret that you created after the workflow run started.

2. Generate a key for reading/writing to that secret-store.
3. Add that [key as a secret to the repository](/actions/security-guides/encrypted-secrets).
4. Create a workflow with:

#### Workflow

> > **Note**
> > `secret-store` refers to an imaginary application with imaginary verbs (`store-secret` and `retrieve-secret`).
> > `some/secret-store@v1` would be responsible for installing that application and configuring it to connect to an `instance` with `credentials`.

Reviewer 3: Otherwise, here's a wording suggestion to add a bit more context:

```suggestion
{% note %}

**Note**: This workflow uses an imaginary secret store, `secret-store`, which has imaginary commands `store-secret` and `retrieve-secret`. `some/secret-store@ 27b31702a0e7fc50959f5ad993c78deac1bdfc29` is an imaginary action that installs the `secret-store` application and configures it to connect to an `instance` with `credentials`.

{% endnote %}
```

> > `some/secret-store@v1` would be responsible for installing that application and configuring it to connect to an `instance` with `credentials`.

Author: I don't have an actual example. We looked at Vault years ago and discarded it. This is the first time in ages where it seemed like a useful thing...

          echo "::add-mask::$the_secret"
{% ifversion actions-save-state-set-output-envs %}
          echo "secret-number=$the_secret" >> "$GITHUB_OUTPUT"
{% else %}
          echo "::set-output name=secret-number::$the_secret"
{% endif %}
      - name: Use that secret output (protected by a mask)
        run: |
{% raw %}
          echo "the secret number is ${{ steps.sets-a-secret.outputs.secret-number }}"
{% endraw %}
```

{% endbash %}

{% powershell %}

```yaml{:copy}
on: push
jobs:
  generate-a-secret-output:
    runs-on: ubuntu-latest
    steps:
      - id: sets-a-secret
        name: Generate, mask, and output a secret
        shell: pwsh
        run: |
          Set-Variable -Name TheSecret -Value (Get-Random)
          Write-Output "::add-mask::$TheSecret"
{% ifversion actions-save-state-set-output-envs %}
          "secret-number=$TheSecret" >> $env:GITHUB_OUTPUT
{% else %}
          Write-Output "::set-output name=secret-number::$TheSecret"
{% endif %}
      - name: Use that secret output (protected by a mask)
        shell: pwsh
        run: |
{% raw %}
          Write-Output "the secret number is ${{ steps.sets-a-secret.outputs.secret-number }}"

> {% endraw %}

Author: <img width="765" alt="image" src="https://user-images.githubusercontent.com/2119212/219968950-fb49bcd3-2464-4b23-86b5-108f011f514b.png">

I don't want to think about the blank lines. I'm sure I could get rid of them by merging everything into a single line, but that's really unreadable.

https://github.com/jsoref/docs-19886-926e50/actions/runs/4217658308/jobs/7321598494
<img width="827" alt="image" src="https://user-images.githubusercontent.com/2119212/219969001-b2f46f77-a55b-4d79-a9f4-b69e4efa5ea3.png">

```
2s
Run Set-Variable -Name TheSecret -Value (Get-Random)
  Set-Variable -Name TheSecret -Value (Get-Random)
  Write-Output "::add-mask::$TheSecret"
  
  Write-Output "secret-number=$TheSecret" >> "$env:GITHUB_OUTPUT"
  shell: /usr/bin/pwsh -command ". '{0}'"
2s
Run Write-Output "the secret number is ***"
  
  Write-Output "the secret number is ***"
  shell: /usr/bin/pwsh -command ". '{0}'"
the secret number is ***
```

>           echo "::add-mask::$the_secret"
> {% ifversion actions-save-state-set-output-envs %}
>           echo "secret-number=$the_secret" >> "$GITHUB_OUTPUT"
> {% else %}
>           echo "::set-output name=secret-number::$the_secret"
> {% endif %}
>       - name: Use that secret output (protected by a mask)
>         run: |
> {% raw %}
>           echo "the secret number is ${{ steps.sets-a-secret.outputs.secret-number }}"
> {% endraw %}
> ```
> 
> {% endbash %}
> 
> {% powershell %}
> 
> ```yaml{:copy}
> on: push
> jobs:
>   generate-a-secret-output:
>     runs-on: ubuntu-latest
>     steps:
>       - id: sets-a-secret
>         name: Generate, mask, and output a secret
>         shell: pwsh
>         run: |
>           Set-Variable -Name TheSecret -Value (Get-Random)
>           Write-Output "::add-mask::$TheSecret"
> {% ifversion actions-save-state-set-output-envs %}
>           "secret-number=$TheSecret" >> $env:GITHUB_OUTPUT
> {% else %}
>           Write-Output "::set-output name=secret-number::$TheSecret"
> {% endif %}
>       - name: Use that secret output (protected by a mask)
>         shell: pwsh
>         run: |
> {% raw %}
>           Write-Output "the secret number is ${{ steps.sets-a-secret.outputs.secret-number }}"
> {% endraw %}

Reviewer 4: Avoid the blank lines in the output by moving the tags to the end of the previous lines:

````suggestion
          echo "::add-mask::$the_secret"{% ifversion actions-save-state-set-output-envs %}
          echo "secret-number=$the_secret" >> "$GITHUB_OUTPUT"{% else %}
          echo "::set-output name=secret-number::$the_secret"{% endif %}
      - name: Use that secret output (protected by a mask)
        run: |{% raw %}
          echo "the secret number is ${{ steps.sets-a-secret.outputs.secret-number }}"{% endraw %}
```

{% endbash %}

{% powershell %}

```yaml{:copy}
on: push
jobs:
  generate-a-secret-output:
    runs-on: ubuntu-latest
    steps:
      - id: sets-a-secret
        name: Generate, mask, and output a secret
        shell: pwsh
        run: |
          Set-Variable -Name TheSecret -Value (Get-Random)
          Write-Output "::add-mask::$TheSecret"{% ifversion actions-save-state-set-output-envs %}
          "secret-number=$TheSecret" >> $env:GITHUB_OUTPUT{% else %}
          Write-Output "::set-output name=secret-number::$TheSecret"{% endif %}
      - name: Use that secret output (protected by a mask)
        shell: pwsh
        run: |{% raw %}
          Write-Output "the secret number is ${{ steps.sets-a-secret.outputs.secret-number }}"{% endraw %}
````

```

{% endpowershell %}

> ### Example: Masking and passing a secret between jobs{% ifversion fpt or ghec or ghes > 3.3 or ghae > 3.3 %} or workflows{% endif %}
> 
> If you want to pass a masked secret between jobs{% ifversion fpt or ghec or ghes > 3.3 or ghae > 3.3 %} or workflows{% endif %}, you should store the secret in a store and then retrieve it in the subsequent job{% ifversion fpt or ghec or ghes > 3.3 or ghae > 3.3 %} or workflow{% endif %}.

Reviewer 4: Version 3.3 is no longer supported, so these tags can be removed:
```suggestion
### Example: Masking and passing a secret between jobs or workflows

If you want to pass a masked secret between jobs or workflows, you should store the secret in a store and then retrieve it in the subsequent job or workflow.
```

#### Setup
1. Set up a secret store to store the secret that you will generate during your workflow. For example, Vault.

> 1. Generate a key for reading and writing to that secret store. Store the key as a repository secret. In the following example workflow, the secret name is `SECRET_STORE_CREDENTIALS`. For more information, see "[Encrypted secrets](/actions/security-guides/encrypted-secrets)."

Reviewer 4: ```suggestion
1. Generate a key for reading and writing to that secret store. Store the key as a repository secret. In the following example workflow, the secret name is `SECRET_STORE_CREDENTIALS`. For more information, see "[AUTOTITLE](/actions/security-guides/encrypted-secrets)."
```

#### Workflow

{% note %}

**Note**: This workflow uses an imaginary secret store, `secret-store`, which has imaginary commands `store-secret` and `retrieve-secret`. `some/secret-store@ 27b31702a0e7fc50959f5ad993c78deac1bdfc29` is an imaginary action that installs the `secret-store` application and configures it to connect to an `instance` with `credentials`.

{% endnote %}

{% bash %}

```yaml{:copy}
on: push

jobs:
  secret-generator:
    runs-on: ubuntu-latest
    steps:
    - uses: some/secret-store@v1

>       with:
> {% raw %}
>         credentials: ${{ secrets.SECRET_STORE_CREDENTIALS }}
>         instance: ${{ secrets.SECRET_STORE_INSTANCE }}
> {% endraw %}
>     - name: generate secret
>       shell: bash
>       run: |
>         GENERATED_SECRET=$((RANDOM))
>         echo "::add-mask::$GENERATED_SECRET"
>         SECRET_HANDLE=$(secret-store store-secret "$GENERATED_SECRET")
> {% ifversion actions-save-state-set-output-envs %}
>         echo "handle=$secret_handle" >> "$GITHUB_OUTPUT"
> {% else %}
>         echo "::set-output name=handle::$secret_handle"
> {% endif %}
>   secret-consumer:
>     runs-on: macos-latest
>     needs: secret-generator
>     steps:
>     - uses: some/secret-store@v1
>       with:
> {% raw %}
>         credentials: ${{ secrets.SECRET_STORE_CREDENTIALS }}
>         instance: ${{ secrets.SECRET_STORE_INSTANCE }}
> {% endraw %}
>     - name: use secret
>       shell: bash
>       run: |
> {% raw %}
>         SECRET_HANDLE="${{ needs.secret-generator.outputs.handle }}"
> {% endraw %}

Reviewer 4: Move tags to avoid blank lines in output:

```suggestion
      with:{% raw %}
        credentials: ${{ secrets.SECRET_STORE_CREDENTIALS }}
        instance: ${{ secrets.SECRET_STORE_INSTANCE }}{% endraw %}
    - name: generate secret
      shell: bash
      run: |
        GENERATED_SECRET=$((RANDOM))
        echo "::add-mask::$GENERATED_SECRET"
        SECRET_HANDLE=$(secret-store store-secret "$GENERATED_SECRET"){% ifversion actions-save-state-set-output-envs %}
        echo "handle=$secret_handle" >> "$GITHUB_OUTPUT"{% else %}
        echo "::set-output name=handle::$secret_handle"{% endif %}
  secret-consumer:
    runs-on: macos-latest
    needs: secret-generator
    steps:
    - uses: some/secret-store@v1
      with:{% raw %}
        credentials: ${{ secrets.SECRET_STORE_CREDENTIALS }}
        instance: ${{ secrets.SECRET_STORE_INSTANCE }}{% endraw %}
    - name: use secret
      shell: bash
      run: |{% raw %}
        SECRET_HANDLE="${{ needs.secret-generator.outputs.handle }}"{% endraw %}
```

        RETRIEVED_SECRET=$(secret-store retrieve-secret "$SECRET_HANDLE")
        echo "::add-mask::$RETRIEVED_SECRET"
        echo "We retrieved our masked secret: $RETRIEVED_SECRET"
```
{% endbash %}

> {% powershell %}

Reviewer 4: There needs to be an empty line after this tag (see problems in the current output: https://docs-19886-926e50.preview.ghdocs.com/en/enterprise-cloud@latest/actions/using-workflows/workflow-commands-for-github-actions?tool=powershell#example-masking-and-passing-a-secret-between-jobs-or-workflows)

```suggestion
{% powershell %}

```

```yaml{:copy}
on: push

jobs:
  secret-generator:
    runs-on: ubuntu-latest
    steps:
    - uses: some/secret-store@v1

>       with:
> {% raw %}
>         credentials: ${{ secrets.SECRET_STORE_CREDENTIALS }}
>         instance: ${{ secrets.SECRET_STORE_INSTANCE }}
> {% endraw %}
>     - name: generate secret
>       shell: pwsh
>       run: |
>         Set-Variable -Name Generated_Secret -Value (Get-Random)
>         Write-Output "::add-mask::$Generated_Secret"
>         Set-Variable -Name Secret_Handle -Value (Store-Secret "$Generated_Secret")
> {% ifversion actions-save-state-set-output-envs %}
>         "handle=$Secret_Handle" >> $env:GITHUB_OUTPUT
> {% else %}
>         Write-Output "::set-output name=handle::$Secret_Handle"
> {% endif %}
>   secret-consumer:
>     runs-on: macos-latest
>     needs: secret-generator
>     steps:
>     - uses: some/secret-store@v1
>       with:
> {% raw %}
>         credentials: ${{ secrets.SECRET_STORE_CREDENTIALS }}
>         instance: ${{ secrets.SECRET_STORE_INSTANCE }}
> {% endraw %}
>     - name: use secret
>       shell: pwsh
>       run: |
> {% raw %}
>         Set-Variable -Name Secret_Handle -Value "${{ needs.secret-generator.outputs.handle }}"
> {% endraw %}

Reviewer 4: ```suggestion
      with:{% raw %}
        credentials: ${{ secrets.SECRET_STORE_CREDENTIALS }}
        instance: ${{ secrets.SECRET_STORE_INSTANCE }}{% endraw %}
    - name: generate secret
      shell: pwsh
      run: |
        Set-Variable -Name Generated_Secret -Value (Get-Random)
        Write-Output "::add-mask::$Generated_Secret"
        Set-Variable -Name Secret_Handle -Value (Store-Secret "$Generated_Secret"){% ifversion actions-save-state-set-output-envs %}
        "handle=$Secret_Handle" >> $env:GITHUB_OUTPUT{% else %}
        Write-Output "::set-output name=handle::$Secret_Handle"{% endif %}
  secret-consumer:
    runs-on: macos-latest
    needs: secret-generator
    steps:
    - uses: some/secret-store@v1
      with:{% raw %}
        credentials: ${{ secrets.SECRET_STORE_CREDENTIALS }}
        instance: ${{ secrets.SECRET_STORE_INSTANCE }}{% endraw %}
    - name: use secret
      shell: pwsh
      run: |{% raw %}
        Set-Variable -Name Secret_Handle -Value "${{ needs.secret-generator.outputs.handle }}"{% endraw %}
```
